File: recognizer.py
"""
recognizer.py

A simplified implementation of the $1 Unistroke Recognizer
(Wobbrock, Wilson & Li, 2007) for classifying a traced path of
(x, y) points into a known shape: circle, square, triangle, or line.

The algorithm:
  1. Resample the stroke into a fixed number of equidistant points.
  2. Rotate so the "indicative angle" (centroid -> first point) is 0.
  3. Scale to a reference bounding box and translate to origin.
  4. Compare against a set of pre-defined template gestures using
     the average point-distance ("Golden Section Search" is used in
     the original paper to find the best rotation alignment; we use
     a simple brute-force angle sweep here for clarity).

This is intentionally dependency-free (just math) so it drops into
any pipeline that hands it a list of points.
"""

import math

NUM_RESAMPLE_POINTS = 64
SQUARE_SIZE = 250.0  # reference bounding box side length used for scaling
ANGLE_SWEEP_RANGE = math.radians(45)
ANGLE_SWEEP_STEP = math.radians(2)


def _path_length(points):
    return sum(
        math.hypot(points[i][0] - points[i - 1][0], points[i][1] - points[i - 1][1])
        for i in range(1, len(points))
    )


def _resample(points, n=NUM_RESAMPLE_POINTS):
    interval = _path_length(points) / (n - 1)
    if interval == 0:
        return [points[0]] * n

    new_points = [points[0]]
    accumulated = 0.0
    pts = list(points)

    i = 1
    while i < len(pts):
        d = math.hypot(pts[i][0] - pts[i - 1][0], pts[i][1] - pts[i - 1][1])
        if accumulated + d >= interval:
            t = (interval - accumulated) / d if d != 0 else 0
            nx = pts[i - 1][0] + t * (pts[i][0] - pts[i - 1][0])
            ny = pts[i - 1][1] + t * (pts[i][1] - pts[i - 1][1])
            new_point = (nx, ny)
            new_points.append(new_point)
            pts.insert(i, new_point)
            accumulated = 0.0
        else:
            accumulated += d
        i += 1

    # Pad in case of float rounding leaving us one short
    while len(new_points) < n:
        new_points.append(pts[-1])

    return new_points[:n]


def _centroid(points):
    cx = sum(p[0] for p in points) / len(points)
    cy = sum(p[1] for p in points) / len(points)
    return cx, cy


def _rotate_to_zero(points):
    cx, cy = _centroid(points)
    theta = math.atan2(points[0][1] - cy, points[0][0] - cx)
    cos_t, sin_t = math.cos(-theta), math.sin(-theta)
    return [
        (
            (p[0] - cx) * cos_t - (p[1] - cy) * sin_t + cx,
            (p[0] - cx) * sin_t + (p[1] - cy) * cos_t + cy,
        )
        for p in points
    ]


def _scale_to_square(points, size=SQUARE_SIZE):
    xs = [p[0] for p in points]
    ys = [p[1] for p in points]
    w, h = max(xs) - min(xs), max(ys) - min(ys)
    w = w if w != 0 else 1e-9
    h = h if h != 0 else 1e-9
    return [((p[0] - min(xs)) * size / w, (p[1] - min(ys)) * size / h) for p in points]


def _translate_to_origin(points):
    cx, cy = _centroid(points)
    return [(p[0] - cx, p[1] - cy) for p in points]


def _rotate_by(points, angle):
    cx, cy = _centroid(points)
    cos_t, sin_t = math.cos(angle), math.sin(angle)
    return [
        (
            (p[0] - cx) * cos_t - (p[1] - cy) * sin_t + cx,
            (p[0] - cx) * sin_t + (p[1] - cy) * cos_t + cy,
        )
        for p in points
    ]


def _path_distance(a, b):
    return sum(math.hypot(a[i][0] - b[i][0], a[i][1] - b[i][1]) for i in range(len(a))) / len(a)


def _best_distance(candidate, template):
    """Brute-force angle sweep to find the best alignment distance."""
    best = float("inf")
    theta = -ANGLE_SWEEP_RANGE
    while theta <= ANGLE_SWEEP_RANGE:
        rotated = _rotate_by(candidate, theta)
        d = _path_distance(rotated, template)
        if d < best:
            best = d
        theta += ANGLE_SWEEP_STEP
    return best


def normalize(points):
    """Runs the full $1 normalization pipeline on a raw list of (x, y) points."""
    resampled = _resample(points)
    rotated = _rotate_to_zero(resampled)
    scaled = _scale_to_square(rotated)
    return _translate_to_origin(scaled)


def _make_template(raw_points):
    return normalize(raw_points)


# --- Reference templates, defined as simple parametric point sets ---

def _square_template():
    pts = []
    for t in range(26):
        pts.append((t / 25 * SQUARE_SIZE, 0))
    for t in range(26):
        pts.append((SQUARE_SIZE, t / 25 * SQUARE_SIZE))
    for t in range(26):
        pts.append((SQUARE_SIZE - t / 25 * SQUARE_SIZE, SQUARE_SIZE))
    for t in range(26):
        pts.append((0, SQUARE_SIZE - t / 25 * SQUARE_SIZE))
    return pts


def _circle_template():
    pts = []
    for i in range(64):
        a = 2 * math.pi * i / 63
        pts.append((SQUARE_SIZE / 2 * math.cos(a), SQUARE_SIZE / 2 * math.sin(a)))
    return pts


def _triangle_template():
    p1 = (SQUARE_SIZE / 2, 0)
    p2 = (0, SQUARE_SIZE)
    p3 = (SQUARE_SIZE, SQUARE_SIZE)
    pts = []
    for t in range(22):
        f = t / 21
        pts.append((p1[0] + f * (p2[0] - p1[0]), p1[1] + f * (p2[1] - p1[1])))
    for t in range(22):
        f = t / 21
        pts.append((p2[0] + f * (p3[0] - p2[0]), p2[1] + f * (p3[1] - p2[1])))
    for t in range(21):
        f = t / 20
        pts.append((p3[0] + f * (p1[0] - p3[0]), p3[1] + f * (p1[1] - p3[1])))
    return pts


def _line_template():
    return [(t / 63 * SQUARE_SIZE, 0) for t in range(64)]


TEMPLATES = {
    "square": _make_template(_square_template()),
    "circle": _make_template(_circle_template()),
    "triangle": _make_template(_triangle_template()),
    "line": _make_template(_line_template()),
}


_DIAGONAL = math.hypot(SQUARE_SIZE, SQUARE_SIZE)
_HALF_DIAGONAL = _DIAGONAL / 2


def recognize(raw_points, min_points=10):
    """
    Classifies a raw stroke (list of (x, y) pixel points) against the
    template shapes. Returns (best_label, score) where score is in
    [0, 1] -- 1.0 is a perfect match. Returns (None, 0.0) if the
    stroke has too few points to be meaningful.
    """
    if len(raw_points) < min_points:
        return None, 0.0

    candidate = normalize(raw_points)

    best_label, best_score = None, -1.0
    for label, template in TEMPLATES.items():
        d = _best_distance(candidate, template)
        score = max(0.0, 1.0 - d / _HALF_DIAGONAL)
        if score > best_score:
            best_label, best_score = label, score

    return best_label, best_score

File: recognizer.js
/**
 * recognizer.js
 *
 * JS port of the Python shape_recognizer.py: a simplified $1
 * Unistroke Recognizer (Wobbrock, Wilson & Li, 2007) that classifies
 * a traced (x, y) point path into square / circle / triangle / line.
 *
 * No dependencies, just math -- resample to a fixed point count,
 * normalize rotation/scale/position, then brute-force sweep a small
 * rotation range to find the best match against each template.
 */

const NUM_RESAMPLE_POINTS = 64;
const SQUARE_SIZE = 250.0;
const ANGLE_SWEEP_RANGE = (45 * Math.PI) / 180;
const ANGLE_SWEEP_STEP = (2 * Math.PI) / 180;

function pathLength(points) {
  let total = 0;
  for (let i = 1; i < points.length; i++) {
    total += Math.hypot(points[i].x - points[i - 1].x, points[i].y - points[i - 1].y);
  }
  return total;
}

function resample(points, n = NUM_RESAMPLE_POINTS) {
  const interval = pathLength(points) / (n - 1);
  if (interval === 0) return new Array(n).fill(points[0]);

  const newPoints = [points[0]];
  let accumulated = 0.0;
  const pts = points.slice();

  let i = 1;
  while (i < pts.length) {
    const d = Math.hypot(pts[i].x - pts[i - 1].x, pts[i].y - pts[i - 1].y);
    if (accumulated + d >= interval) {
      const t = d !== 0 ? (interval - accumulated) / d : 0;
      const newPoint = {
        x: pts[i - 1].x + t * (pts[i].x - pts[i - 1].x),
        y: pts[i - 1].y + t * (pts[i].y - pts[i - 1].y),
      };
      newPoints.push(newPoint);
      pts.splice(i, 0, newPoint);
      accumulated = 0.0;
    } else {
      accumulated += d;
    }
    i += 1;
  }

  while (newPoints.length < n) newPoints.push(pts[pts.length - 1]);
  return newPoints.slice(0, n);
}

function centroid(points) {
  const cx = points.reduce((s, p) => s + p.x, 0) / points.length;
  const cy = points.reduce((s, p) => s + p.y, 0) / points.length;
  return { x: cx, y: cy };
}

function rotateToZero(points) {
  const c = centroid(points);
  const theta = Math.atan2(points[0].y - c.y, points[0].x - c.x);
  return rotateBy(points, -theta, c);
}

function rotateBy(points, angle, aboutPoint = null) {
  const c = aboutPoint || centroid(points);
  const cosT = Math.cos(angle);
  const sinT = Math.sin(angle);
  return points.map((p) => ({
    x: (p.x - c.x) * cosT - (p.y - c.y) * sinT + c.x,
    y: (p.x - c.x) * sinT + (p.y - c.y) * cosT + c.y,
  }));
}

function scaleToSquare(points, size = SQUARE_SIZE) {
  const xs = points.map((p) => p.x);
  const ys = points.map((p) => p.y);
  const minX = Math.min(...xs), maxX = Math.max(...xs);
  const minY = Math.min(...ys), maxY = Math.max(...ys);
  const w = maxX - minX || 1e-9;
  const h = maxY - minY || 1e-9;
  return points.map((p) => ({
    x: ((p.x - minX) * size) / w,
    y: ((p.y - minY) * size) / h,
  }));
}

function translateToOrigin(points) {
  const c = centroid(points);
  return points.map((p) => ({ x: p.x - c.x, y: p.y - c.y }));
}

function pathDistance(a, b) {
  let total = 0;
  for (let i = 0; i < a.length; i++) {
    total += Math.hypot(a[i].x - b[i].x, a[i].y - b[i].y);
  }
  return total / a.length;
}

function bestDistance(candidate, template) {
  let best = Infinity;
  for (let theta = -ANGLE_SWEEP_RANGE; theta <= ANGLE_SWEEP_RANGE; theta += ANGLE_SWEEP_STEP) {
    const rotated = rotateBy(candidate, theta);
    const d = pathDistance(rotated, template);
    if (d < best) best = d;
  }
  return best;
}

function normalize(points) {
  const resampled = resample(points);
  const rotated = rotateToZero(resampled);
  const scaled = scaleToSquare(rotated);
  return translateToOrigin(scaled);
}

// ----------------------------------------------------- reference templates

function squareTemplateRaw() {
  const pts = [];
  for (let t = 0; t <= 25; t++) pts.push({ x: (t / 25) * SQUARE_SIZE, y: 0 });
  for (let t = 0; t <= 25; t++) pts.push({ x: SQUARE_SIZE, y: (t / 25) * SQUARE_SIZE });
  for (let t = 0; t <= 25; t++) pts.push({ x: SQUARE_SIZE - (t / 25) * SQUARE_SIZE, y: SQUARE_SIZE });
  for (let t = 0; t <= 25; t++) pts.push({ x: 0, y: SQUARE_SIZE - (t / 25) * SQUARE_SIZE });
  return pts;
}

function circleTemplateRaw() {
  const pts = [];
  for (let i = 0; i < 64; i++) {
    const a = (2 * Math.PI * i) / 63;
    pts.push({ x: (SQUARE_SIZE / 2) * Math.cos(a), y: (SQUARE_SIZE / 2) * Math.sin(a) });
  }
  return pts;
}

function triangleTemplateRaw() {
  const p1 = { x: SQUARE_SIZE / 2, y: 0 };
  const p2 = { x: 0, y: SQUARE_SIZE };
  const p3 = { x: SQUARE_SIZE, y: SQUARE_SIZE };
  const pts = [];
  for (let t = 0; t <= 21; t++) {
    const f = t / 21;
    pts.push({ x: p1.x + f * (p2.x - p1.x), y: p1.y + f * (p2.y - p1.y) });
  }
  for (let t = 0; t <= 21; t++) {
    const f = t / 21;
    pts.push({ x: p2.x + f * (p3.x - p2.x), y: p2.y + f * (p3.y - p2.y) });
  }
  for (let t = 0; t <= 20; t++) {
    const f = t / 20;
    pts.push({ x: p3.x + f * (p1.x - p3.x), y: p3.y + f * (p1.y - p3.y) });
  }
  return pts;
}

function lineTemplateRaw() {
  const pts = [];
  for (let t = 0; t < 64; t++) pts.push({ x: (t / 63) * SQUARE_SIZE, y: 0 });
  return pts;
}

const TEMPLATES = {
  square: normalize(squareTemplateRaw()),
  circle: normalize(circleTemplateRaw()),
  triangle: normalize(triangleTemplateRaw()),
  line: normalize(lineTemplateRaw()),
};

const DIAGONAL = Math.hypot(SQUARE_SIZE, SQUARE_SIZE);
const HALF_DIAGONAL = DIAGONAL / 2;

/**
 * Classifies a raw stroke (array of {x, y} pixel points) against the
 * template shapes. Returns { label, score } where score is in [0, 1]
 * (1.0 = perfect match). Returns { label: null, score: 0 } if the
 * stroke has too few points.
 */
function recognizeShape(rawPoints, minPoints = 10) {
  if (rawPoints.length < minPoints) return { label: null, score: 0.0 };

  const candidate = normalize(rawPoints);

  let bestLabel = null;
  let bestScore = -1.0;
  for (const [label, template] of Object.entries(TEMPLATES)) {
    const d = bestDistance(candidate, template);
    const score = Math.max(0.0, 1.0 - d / HALF_DIAGONAL);
    if (score > bestScore) {
      bestLabel = label;
      bestScore = score;
    }
  }

  return { label: bestLabel, score: bestScore };
}
